### ISSUE TITLE
Add UInt32 and UInt64 indexers

### ISSUE THREAD

**Issue by AlexDBlack** [bytedeco/javacpp#376]
Motivation: https://github.com/eclipse/deeplearning4j/issues/8582

The main challenges here are of course the numerical ranges vs. java types, in terms of API for getting/setting values.
uint32 isn't so bad, as it fits in a long (like UByteIndexer uses int)
However uint64 can clearly represent values larger than long.

Not sure what the best solution for uint64 is, but some options are:
* use long (ignoring the overflow)
* add a new/custom Number class (uint64 stored as a long, converted as required)
* BigInt (not exactly a low overhead solution however)

**Comment by saudet** [comment]
We can still use `int` and `long`. Java SE provides unsigned operations for them: https://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html
Those are helper methods instead of wrapper classes for efficiency.

**Comment by AlexDBlack** [comment]
I wasn't aware of the unsigned operations on Integer and Long - some are Java 8, like Long.toUnsignedString, but not all.


OK, if we go with "unsigned as signed" (i.e., use int/long as the storage format) I guess there's some design options here...

1. Don't add new indexers, and have users just use the existing (signed) int32/64 indexers
i.e., user is responsible for remembering that 
That's clearly the fastest/easiest option, but that seems error prone, for a couple of reasons:
(a) the "int/long that should be interpreted as unsigned" vs. "real" int/long values (easy to mix up which is which)
(b) getDouble doesn't work (hence Indexer.toString shows wrong values)

2. Add new uint32/64 indexers, but only really change getDouble - otherwise pretty much the same API as the existing Int/Long(Raw)Indexer classes

3. Add new uint32/64 indexers, but have a bunch of overloads for the "real" vs. "storage" int/long cases.
By that I mean the API (for UInt64Indexer) would have methods like:
```
long get(...)                                   //Returns uint64 stored as a long
long getSigned(...)                             //Converts unsigned to signed (returns a "real" long), can overflow
UInt64Indexer put(long idx, long value)         //Put a uint64 stored as a long
UInt64Indexer putSigned(long idx, long value)   //Put a "real" long value
```


The UInt32Indexer would look like:
```
int get(...)                                    //Returns uint32 stored as an int
long getSigned(...)                             //Converts unsigned to signed
UInt32Indexer put(long idx, int value)          //Put a uint32 stored as a int
UInt32Indexer putSigned(long idx, long value)   //Put a "real" uint32 value (stored as a long)
```


Option 3 makes the most sense to me - if that looks OK @Saudet, I can look at doing a PR with that approach

**Comment by saudet** [comment]
Unsigned versions of indexers like that will just be a copy/paste of the
generic ones. I'm not sure I understand how useful that is going to be.
Maybe what you mean is adding unsigned operations to the generic ones as is
done for HalfIndexer? That would make sense, yes.

**Comment by saudet** [comment]
I've added `UIntIndexer` that does something similar to `UByteIndexer` and `UShortIndexer`, but widens the values to `long` instead of `int`, which makes sense since most real machines are 64-bit these days. But we won't be able do the same for a `ULongIndexer` where, for example, the implementations of your `putSigned()` and `getSigned()` methods would be exactly the same as the `put()` and `get()` methods from `LongIndexer`.

**Comment by AlexDBlack** [comment]
Thanks, new UInt indexers looks good, and should do what I need here

As for ULongIndexer - right, I'm not seeing a good solution here
If we stick with primitives (and 'ulong as long'), the only real advantage to a ULongIndexer is fixing toString and adding some javadoc to aid usability. Maybe not worth it just for that.

So unless we can come up with a better idea, maybe I'll try treating the LongIndexer+UINT64 in ND4J as a special case and handle the conversion there where necessary, and we can close this.

**Comment by saudet** [comment]
We could add a `LongIndexer.toUnsignedString()` to mimic `Long.toUnsignedString()`? Similarly for `getUnsignedDouble()` and `putUnsignedDouble()`. That would be it I think? Having a class named "unsigned" not actually doing unsigned anything is misleading, but if it's only a couple of methods that actually do unsigned operations, that sounds fine.

**Comment by AlexDBlack** [comment]
> We could add a LongIndexer.toUnsignedString() to mimic Long.toUnsignedString()? Similarly for getUnsignedDouble() and putUnsignedDouble().

Sure, that sounds good to me.

**Comment by saudet** [comment]
I'm thinking we could have a ULongIndexer that uses BigInteger just to have something that has easy-to-use operations without needing some custom class that wouldn't be that fast anyway, while for performance we can still always use LongIndexer.

**Comment by AlexDBlack** [comment]
Right, I proposed using BigInteger as one of the options earlier.
That might be the best compromise between performance and usability - i.e., provide both options and let users choose.
I might stick with LongIndexer in ND4J (for performance reasons), but a BigInteger-based ULongIndexer could definitely be useful for other use cases.

**Comment by saudet** [comment]
New indexers are included in just released version 1.5.3. Thanks for the feedback!
